src/clj/puppetlabs/services/legacy_routes/legacy_routes_core.clj

```diff
@@ -0,0 +1,144 @@
+              (dissoc :params :route-params))]
+      (handler compat-request))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Routing
+
+(defn v2_0-routes
+  "Creates the compojure routes to handle the 3.X master's '/v2.0' routes."
+  [request-handler]
+  (compojure/routes
+    (compojure/GET "/environments" request
+                   (request-handler request))))
+
+(defn legacy-master-routes
+  [master-request-handler ca-request-handler]
+  (compojure/routes
+    (compojure/GET "/node/*" request (master-request-handler request))
+    (compojure/GET "/facts/*" request (master-request-handler request))
+    (compojure/GET "/file_content/*" request (master-request-handler request))
+    (compojure/GET "/file_metadatas/*" request (master-request-handler request))
+    (compojure/GET "/file_metadata/*" request (master-request-handler request))
+    (compojure/GET "/file_bucket_file/*" request
+      (master-request-handler (map-accept-s-pson-to-binary request)))
+    ;; TODO: file_bucket_file request PUTs from Puppet agents currently use a
+    ;; Content-Type of 'text/plain', which, per HTTP specification, would imply
+    ;; a default character encoding of ISO-8859-1 or US-ASCII be used to decode
+    ;; the data.  This would be incorrect to do in this case, however, because
+    ;; the actual payload is "binary".  Coercing this to
+    ;; "application/octet-stream" for now as this is synonymous with "binary".
+    ;; This should be removed when/if Puppet agents start using an appropriate
+    ;; Content-Type to describe the input payload - see PUP-3812 for the core
+    ;; Puppet work and SERVER-294 for the related Puppet Server work that
+    ;; would be done.
+    (compojure/PUT "/file_bucket_file/*" request
+      (master-request-handler
+        (-> request
+          (map-accept-s-pson-to-binary)
+          (assoc :content-type "application/octet-stream") ; deprecated in ring >= 1.2
+          (assoc-in [:headers "content-type"] "application/octet-stream")))) ; ring >= 1.2 spec
+    (compojure/HEAD "/file_bucket_file/*" request
+      (master-request-handler (map-accept-s-pson-to-binary request)))
+    (compojure/GET "/catalog/*" request (master-request-handler request))
+    (compojure/POST "/catalog/*" request (master-request-handler request))
+    (compojure/PUT "/report/*" request (master-request-handler request))
+    (compojure/GET "/resource_type/*" request (master-request-handler request))
+    (compojure/GET "/resource_types/*" request (master-request-handler request))
+    (compojure/GET "/status/*" request (master-request-handler request))
+
+    ;; TODO: when we get rid of the legacy dashboard after 3.4, we should remove
+    ;; this endpoint as well.  It makes more sense for this type of query to be
+    ;; directed to PuppetDB.
+    (compojure/GET "/facts_search/*" request (master-request-handler request))
```

Comment:
@cprice404 or @jonathannewman - Do you know if the next release of PE - the first that would include Puppet Server 2 / Puppet 4 support - would be one in which the "facts_search" endpoint would no longer need to be used by the PE UI?  Since this PR is targeted for work to be used in that timeframe, I wonder if we can have this PR discontinue support for this endpoint.